### PR TITLE
build: split pthread flags out of ldflags and dont use when building libconsensus

### PR DESCRIPTION
**Pull request by fanquake**
TLDR: Split pthread flags out of ldflags, and stop using them when building libconsensus.

Building libconsensus on Linux using Clang currently warns. i.e:
```bash
./autogen.sh
./configure --disable-tests --disable-bench --with-utils=no --with-daemon=no --with-gui=no --disable-wallet --with-libs=yes CC=clang CXX=clang++
make V=1 -j6
... -Wl,-z -Wl,relro -Wl,-z -Wl,now   -pthread -Wl,-soname -Wl,libbitcoinconsensus.so.0 -o .libs/libbitcoinconsensus.so.0.0.0
clang: warning: argument unused during compilation: '-pthread' [-Wunused-command-line-argument]
clang: warning: argument unused during compilation: '-pthread' [-Wunused-command-line-argument]
```

Besides wanting to quiet the warnings, after digging into this it seemed we could clean up how we are passing around the pthread flags. I also learnt a bit more about how libtools builds shared libraries, and that passing `-pthread` on the link line wouldn't be enough to link against pthreads anyways, due to libtools usage of -nostdlib (see [related discussion where we build DLLs](https://github.com/bitcoin/bitcoin/blob/476436b2dec254bb988f8c7a6cbec1d7bb7cecfd/configure.ac#L603)).

This can be demonstrated with a patch to libconsensus:
```patch
diff --git a/src/script/bitcoinconsensus.cpp b/src/script/bitcoinconsensus.cpp
index 15e204062..10bf3582f 100644
--- a/src/script/bitcoinconsensus.cpp
+++ b/src/script/bitcoinconsensus.cpp
@@ -10,6 +10,8 @@
 #include <script/interpreter.h>
 #include <version.h>
 
+#include <pthread.h>
+
 namespace {
 
 /** A class that deserializes a single CTransaction one time. */
@@ -127,3 +129,10 @@ unsigned int bitcoinconsensus_version()
     // Just use the API version for now
     return BITCOINCONSENSUS_API_VER;
 }
+
+void *func_pthread(void *x) { return x; }
+
+void f() {
+	pthread_t t;
+	pthread_create(&t,0,func_pthread,0);
+}
```

After building,  you'll find you have a `libbitcoinconsensus.so` using pthread symbols, but which isn't linked against libpthread:
```bash
ldd -r src/.libs/libbitcoinconsensus.so
	linux-vdso.so.1 (0x00007ffe49378000)
	libstdc++.so.6 => /usr/lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007f553cee7000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f553cda2000)
	libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f553cd88000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f553cbc5000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f553d15d000)
undefined symbol: pthread_create	(src/.libs/libbitcoinconsensus.so)
```

This libtool behaviour has been known about for some time, i.e this [thread from 2005](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=25460),  describes the same issue. The suggestion from libtool maintainers at the time is to add `-lpthread` to LDFLAGS.

Also worth noting is that some of the users in those threads were also using the `AX_PTHREADS` macro, same as us, to determine how to compile with/link against pthreads. This macro has [recently been updated](https://git.savannah.gnu.org/gitweb/?p=autoconf-archive.git;a=commitdiff;h=2fb904589643eb6ca6122f834891b58d1d51b347), with reference to this issue. You can compare the output from the version we currently use, to the new version:
```bash
# our ax_pthread macro:
  PTHREAD_CFLAGS = -pthread
  PTHREAD_LIBS  = 
  PTHREAD_CC    = gcc / clang

# the new ax_pthread macro
  PTHREAD_CFLAGS = -pthread
  PTHREAD_LIBS  = -lpthread
  PTHREAD_CC    = gcc / clang
```

Note that as part of this PR I've also added `PTHREAD_LIBS` to the split out flags. Although we weren't using it anywhere previously (and wouldn't have seemed to matter for the most part, given it was likely empty for most builders), the macro assumes it's use. i.e:
> NOTE: You are assumed to not only compile your program with these flags,
> but also to link with them as well. For example, you might link with
> $PTHREAD_CC $CFLAGS $PTHREAD_CFLAGS $LDFLAGS ... $PTHREAD_LIBS $LIBS

### CODE REVIEW

File: configure.ac [hunk 1]
```diff
@@ -1703,11 +1703,13 @@ echo
...
-echo "  LDFLAGS       = $PTHREAD_CFLAGS $HARDENED_LDFLAGS $GPROF_LDFLAGS $LDFLAGS"
-echo "  ARFLAGS       = $ARFLAGS"
+echo "  CC             = $CC"
+echo "  CFLAGS         = $CFLAGS"
+echo "  CPPFLAGS       = $DEBUG_CPPFLAGS $HARDENED_CPPFLAGS $CPPFLAGS"
+echo "  CXX            = $CXX"
+echo "  CXXFLAGS       = $DEBUG_CXXFLAGS $HARDENED_CXXFLAGS $WARN_CXXFLAGS $NOWARN_CXXFLAGS $ERROR_CXXFLAGS $GPROF_CXXFLAGS $CXXFLAGS"
+echo "  LDFLAGS        = $HARDENED_LDFLAGS $GPROF_LDFLAGS $LDFLAGS"
+echo "  ARFLAGS        = $ARFLAGS"
+echo "  PTHREAD_CFLAGS = $PTHREAD_CFLAGS"
```
Comment (laanwj) [inline configure.ac:1713]: I'm not sure listing these separately here is worth it. We don't do so dor other libraries, and don't want this list to get too long.

**Comment by laanwj** [comment]
Concept ACK. I don't think there are plans to ever make `libconsensus` use threads (definitely not directly), because this would make it harder to use from some languages. So this seems fine.

File: configure.ac (hunk 1 above)
Comment (fanquake) [reply configure.ac]: Fair enough. I've removed split out.

**Comment by laanwj** [comment]
@dongcarl @theuni can you take a look here please

**Comment by dongcarl** [comment]
Looks like the builds are working.
A quick skim of the code looks correct, however, I willingly admit that I'm not the most familiar with automake idiosyncrasies and conventions.

**Review by hebasto — APPROVED** [review]
ACK 9c5bef450f7bf4dd4307cae8bce76dc131643574, tested on Linux Mint 20 (x86_64) -- no clang warnings now. Verified that `ax_pthread.m4` has no our custom patches.

Mind making 8331ad7538e1e3c6efc98fc8fe5662c3d6d953b6 a scripted-diff:
```
-BEGIN VERIFY SCRIPT-
sed -i -e 's/\$(RELDFLAGS) \$(AM_LDFLAGS) \$(LIBTOOL_APP_LDFLAGS)$/\$(FUZZ_SUITE_LDFLAGS_COMMON)/' src/Makefile.test.include
patch -p1 << "EOF"
--- a/src/Makefile.test.include
+++ b/src/Makefile.test.include
@@ -318,6 +318,8 @@ endif
 
 if ENABLE_FUZZ
 
+FUZZ_SUITE_LDFLAGS_COMMON = $(RELDFLAGS) $(AM_LDFLAGS) $(LIBTOOL_APP_LDFLAGS)
+
 test_fuzz_addition_overflow_CPPFLAGS = $(AM_CPPFLAGS) $(BITCOIN_INCLUDES)
 test_fuzz_addition_overflow_CXXFLAGS = $(AM_CXXFLAGS) $(PIE_FLAGS)
 test_fuzz_addition_overflow_LDADD = $(FUZZ_SUITE_LD_COMMON)
EOF
-END VERIFY SCRIPT-
```
?

**Comment by fanquake** [comment]
> Mind making 8331ad7 a scripted-diff:

It's now a scripted diff.

**Review by hebasto — APPROVED** [review]
re-ACK d42f2150496128fde0a9d2d28707a000a7a0d900
```
$ git range-diff master 9c5bef450f7bf4dd4307cae8bce76dc131643574 d42f2150496128fde0a9d2d28707a000a7a0d900
1:  ae409c1b4 = 1:  3711cb7ce build: add PTHREAD_LIBS to LDFLAGS configure output
2:  8331ad753 ! 2:  b3ee267a6 test: add FUZZ_SUITE_LDFLAGS_COMMON
    @@ Metadata
     Author: fanquake <fanquake@gmail.com>
     
      ## Commit message ##
    -    test: add FUZZ_SUITE_LDFLAGS_COMMON
    +    scripted-diff: add FUZZ_SUITE_LDFLAGS_COMMON
    +
    +    -BEGIN VERIFY SCRIPT-
    +    sed -i -e 's/\$(RELDFLAGS) \$(AM_LDFLAGS) \$(LIBTOOL_APP_LDFLAGS)$/\$(FUZZ_SUITE_LDFLAGS_COMMON)/' src/Makefile.test.include
    +    patch -p1 << "EOF"
    +    --- a/src/Makefile.test.include
    +    +++ b/src/Makefile.test.include
    +    @@ -318,6 +318,8 @@ endif
    +
    +     if ENABLE_FUZZ
    +
    +    +FUZZ_SUITE_LDFLAGS_COMMON = $(RELDFLAGS) $(AM_LDFLAGS) $(LIBTOOL_APP_LDFLAGS)
    +    +
    +     test_fuzz_addition_overflow_CPPFLAGS = $(AM_CPPFLAGS) $(BITCOIN_INCLUDES)
    +     test_fuzz_addition_overflow_CXXFLAGS = $(AM_CXXFLAGS) $(PIE_FLAGS)
    +     test_fuzz_addition_overflow_LDADD = $(FUZZ_SUITE_LD_COMMON)
    +    EOF
    +    -END VERIFY SCRIPT-
     
      ## src/Makefile.test.include ##
     @@ src/Makefile.test.include: endif
3:  06e0771ec = 3:  abdd60af7 build: split PTHREAD_* flags out of AM_LDFLAGS
4:  9c5bef450 = 4:  d42f21504 build: AX_PTHREAD serial 27
```

File: src/Makefile.am [hunk 2]
```diff
@@ -4,9 +4,10 @@
 
 DIST_SUBDIRS = secp256k1 univalue
 
-AM_LDFLAGS = $(PTHREAD_CFLAGS) $(LIBTOOL_LDFLAGS) $(HARDENED_LDFLAGS) $(GPROF_LDFLAGS) $(SANITIZER_LDFLAGS)
+AM_LDFLAGS = $(LIBTOOL_LDFLAGS) $(HARDENED_LDFLAGS) $(GPROF_LDFLAGS) $(SANITIZER_LDFLAGS)
 AM_CXXFLAGS = $(DEBUG_CXXFLAGS) $(HARDENED_CXXFLAGS) $(WARN_CXXFLAGS) $(NOWARN_CXXFLAGS) $(ERROR_CXXFLAGS) $(GPROF_CXXFLAGS) $(SANITIZER_CXXFLAGS)
 AM_CPPFLAGS = $(DEBUG_CPPFLAGS) $(HARDENED_CPPFLAGS)
+AM_PTHREAD_FLAGS = $(PTHREAD_CFLAGS) $(PTHREAD_LIBS)
```
Comment (theuni) [inline src/Makefile.am:10]: AM_FOO usually have some meaning to Automake, so this should probably drop the AM_ prefix.

**Comment by theuni** [comment]
@hebasto Good call on the scripted diff. That removes a big hurdle to this change.

Pretty sure we're still not really using the pthreads macro as intended, but our usage has held up in the real-world so far, and I don't think this is likely to break anything.

Concept ACK.

File: src/Makefile.am (hunk 2 above)
Comment (fanquake) [reply src/Makefile.am]: Good call. Switch to `PTHREAD_FLAGS`.

**Comment by hebasto** [comment]
It seems in 69a933ca29af37f6568d5faa07cb33c4cffed602 the patched lines location should be updated.

**Comment by fanquake** [comment]
> It seems in 69a933c the patched lines location should be updated.

Thanks. This should be addressed now.

**Review by hebasto — APPROVED** [review]
re-ACK fc9278d162c342bace8a147da6bc4f9941d8d9d7, only rebased and renamed s/`AM_PTHREAD_FLAGS`/`PTHREAD_FLAGS`/ since my [previous](https://github.com/bitcoin/bitcoin/pull/19558#pullrequestreview-473487730) review..

**Review by kallewoof — COMMENTED** [review]
ACK fc9278d162c342bace8a147da6bc4f9941d8d9d7

Code review + compiled and ran tests on mac & linux.

**Comment by laanwj** [comment]
Looks like a straightforward improvement.
Code review ACK fc9278d162c342bace8a147da6bc4f9941d8d9d7
